### ISSUE TITLE
Proxy and restore recursively before assignment

### ISSUE THREAD

**Issue by lionel-** [r-lib/vctrs#1107]
To fix premature restoration. Will help with #1098.

**Comment by lionel-** [comment]
Proxy and restore at first level in `vec_rbind()` as a first step:

```r
evalq({
  local_methods(
    vec_restore.vctrs_foobar = function(x, to, ...) {
      if (anyDuplicated(x$x, incomparables = NA)) {
        abort("tilt")
      }
      foobar(x)
    }
  )

  df <- foobar(data.frame(x = 1))
  vec_rbind(df, df)
})
#>   x
#> 1 1
#> 2 1
```

**Comment by lionel-** [comment]
Will solve the proxy-assign conundrum where input is proxied but value is not.

**Comment by lionel-** [comment]
Reconsider whether data frames can be proxied (without the vcols implemented in #730).

**Comment by lionel-** [comment]
Here is how we could support all of these features simultaneously:

- List proxying into data frames to support vectorised attributes

- Data frames proxying into different data frames (typically extra columns for vectorised attributes)

- Recursive proxying and restoration

- Restore methods that inspect the types and/or the contents of their columns (which seems common in tidymodels)

- Backward compatibility with existing `vec_restore()` methods


What we gain:

1. Supporting extra columns in df-proxies will be useful to implement vectorised attributes.

2. Supporting recursive proxying and restoration will be useful to simplify the internals of vctrs, fix the semantics of restoration (restore methods are currently called on partially assembled data when there are df-cols), and improve performance (no excessive restoration, row-binding in parallel).


How to do this:

1. We proxy recursively over data frame columns. Proxying might create more df-cols in the proxy if there are any record vectors (POSIXlt, vctrs_vctr, ...). For this reason the recursive proxying goes inside out, from innermost df-cols to the outermost data frame. This way we don't proxy again newly created columns.

2. When the transformation is done and it's time to restore, we first "unproxy" all data frames, following df-cols recursively from the outermost data frame to the innermost ones. For record vectors, the unproxying eliminates the df-col and recursion ends there. For data frames, the unproxying restores its overall structure (names and length, i.e. the identity of each column).

3. When rolling back the recursion, we call `vec_restore()` on each node, restoring from innermost columns to the outermost data frame. Because we restore inside out, the restore methods are called with fully restored columns. Their types and contents can be checked during restoration, if needed.

    Each time we recursed unproxying/restoration with a new column of the proxy, we passed the corresponding column of the target prototype. This is only possible if the columns match. That's why it's crucial that `vec_unproxy()` restores the structure (the names and identities of the columns). In order to call `vec_restore()` recursively, we must be able to match by name each restored column to the corresponding column in the target prototype.



Very few classes will need to know about `vec_unproxy()` and the new mechanism should be backward compatible:

- It is currently not supported to add/change columns in the proxy of data frames. All the restoration code for data frames is contained in `vec_restore()` methods that will keep being called as appropriate.

- `vec_unproxy.vctrs_rcrd()` and `vec_unproxy.POSIXlt()` simply unclass their inputs so they become simple lists. This ends the unproxy recursion here and `vec_restore()` is called on the list, allowing the restore methods of  `vctrs_rcrd` subclasses to perform restoration as usual. Unless they expect a data frame in return (they normally shouldn't), this will be backward compatible.


This should get us most of the way there to support extra columns in df proxies. However we'll still have the `vec_cbind()` conundrum to solve. The current workaround of devolving dfs with vectorised attributes into a simple data frame or tibble (by implementing `vec_cbind_frame_ptype()` specially) is always possible though.
